// pang/node_modules/Game.js
var Player = require('./Player.js'),
    MAX_PLAYERS = 2;

function Game (id) {
  this.id = id;
  this.players = Object.create(null);
}

Game.prototype.join = function (socket) {
  var player = new Player(socket, this);
  if (!this.players[socket.id]) {
    this.players[socket.id] = player;
  }

  this.broadcast("player_joined");
};

Game.prototype.leave = function (socket) {

};

Game.prototype.broadcast = function (event, data, socket) {
  var that = this;
  Object.keys(this.players).forEach(function (key) {
    var player = that.players[key];
    // If socket is specified, don't broadcast to this socket.
    if (socket && key === socket.id) {
      return;
    }

    player.socket.emit(event, data);
  });
};

Game.prototype.playerCount = function () {
  return Object.keys(this.players).length;
};

Game.prototype.isFull = function () {
  return this.playerCount() == MAX_PLAYERS;
};

Game.prototype.isAllPlayersReady = function () {
  var that = this;
  return Object.keys(this.players).every(function (key) {
    return that.players[key].ready;
  });
};


module.exports = Game;
