// pang/node_modules/Player.js

function Player (socket, game) {
  this.socket = socket;
  this.game = game;
  this.ready = false;
  this.offset = {
    x: 10,
    y: 250
  };
  this.height = 100;
  this.width = 6;
  this.score = 0;

  this.moving = false;

  if(Object.keys(game.players).length < 1) {
    this.side = 'left';
  } else {
    var other_player_side = game.players[Object.keys(game.players)[0]].side;
    this.side = other_player_side === 'left' ? 'right' : 'left';
  }
  //this.side = Object.keys(this.game.players).length ? 'right' : 'left';
  this.events();
}

Player.prototype.move = function (direction) {
  if (direction === 'up') {
    if(this.offset.y > 0) {
      this.offset.y -= 4;
    }
  } else if (direction === 'down') {
    if(this.offset.y < 500) {
      this.offset.y += 4;  
    }
  }
  
  this.game.broadcast("move", {offset : this.offset, side : this.side});
}

Player.prototype.events = function () {
  var that = this;
  this.socket.on('ready', function () {
    that.ready = true;

    that.game.broadcast("ready", that.side);
  });

  // this.socket.on('score', function() {
  //   that.game.gameStarted = false;
  // });

  this.socket.on('move', function (data) {
    if(!that.moving) {
      that.moving = true;
      setTimeout(function() {
        that.move(data.direction);
        that.moving = false;
      }, 10);
    }
    
    
  });
}

module.exports = Player;
